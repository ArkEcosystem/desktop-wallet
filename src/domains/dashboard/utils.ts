import { Profile, ProfileSetting } from "@arkecosystem/platform-sdk-profiles";
import { uniq } from "@arkecosystem/utils";

export const enableNetworkInDashboardFilters = (profile: Profile, networkId: string) => {
	const config: any = profile.settings().get(ProfileSetting.DashboardConfiguration) || { selectedNetworkIds: [] };
	config.selectedNetworkIds.push(networkId);
	profile.settings().set(ProfileSetting.DashboardConfiguration, {
		...config,
		selectedNetworkIds: uniq(config.selectedNetworkIds),
	});
};
