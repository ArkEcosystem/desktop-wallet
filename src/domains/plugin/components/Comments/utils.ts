import { DateTime } from "@arkecosystem/platform-sdk-intl";

export const getDateDifferenceFromNow = (targetDate: string): string => {
	const currentDateObj = DateTime.make();
	const dateToCompare = DateTime.make(targetDate);
	const dateDifference = currentDateObj.diffInDays(dateToCompare);

	const dateComplement = dateDifference > 1 ? "days" : "day";

	return `${dateDifference} ${dateComplement} ago`;
};
