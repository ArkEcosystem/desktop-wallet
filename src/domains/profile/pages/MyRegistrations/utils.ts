import { DelegateService, ExtendedTransactionData, ReadWriteWallet } from "@arkecosystem/platform-sdk-profiles";

export const filterEntities = (entities: ExtendedTransactionData[], query: string) => {
	if (!query || query === "") return entities;

	const matchesAssetName = (entity: ExtendedTransactionData, query: string) => {
		const { data }: any = entity.asset();
		return data.name.toLowerCase().includes(query.toLowerCase());
	};

	const matchesAddress = (entity: ExtendedTransactionData, query: string) =>
		entity.wallet().address().toLowerCase().includes(query.toLowerCase());

	const matchesWalletName = (entity: ExtendedTransactionData, query: string) =>
		entity.wallet().alias()?.toLowerCase().includes(query.toLowerCase());

	return entities.filter(
		(e) => matchesAssetName(e, query) || matchesWalletName(e, query) || matchesAddress(e, query),
	);
};

export const filterDelegates = (delegateWallets: ReadWriteWallet[], delegates: DelegateService, query: string) => {
	const matchesUserName = (wallet: ReadWriteWallet, query: string) => {
		const delegateInfo = delegates.findByAddress(wallet.coinId(), wallet.networkId(), wallet.address());
		return delegateInfo.username()?.toLowerCase().includes(query);
	};

	return delegateWallets.filter((e) => matchesUserName(e, query));
};
