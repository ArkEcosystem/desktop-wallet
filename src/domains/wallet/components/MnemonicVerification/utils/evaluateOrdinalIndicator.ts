const ordinalPluralizer = new Intl.PluralRules("en", { type: "ordinal" });

const SUFFIXES = {
	zero: "",
	one: "st",
	two: "nd",
	few: "rd",
	other: "th",
	many: "th",
};

export const getOrdinalIndicator = (number: number) => SUFFIXES[ordinalPluralizer.select(number)];
