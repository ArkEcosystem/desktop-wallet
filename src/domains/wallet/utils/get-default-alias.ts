import { Contracts } from "@arkecosystem/platform-sdk-profiles";

interface GetDefaultAliasInput {
	profile: Contracts.IProfile;
	ticker: string;
}

export const getDefaultAlias = ({ profile, ticker }: GetDefaultAliasInput): string => {
	const makeAlias = (count: number) => `${ticker} #${count}`;

	let sameCoinWalletsCount = 1;

	const sameCoinWallets = profile.wallets().allByCoin()[ticker];

	if (sameCoinWallets) {
		sameCoinWalletsCount = Object.keys(sameCoinWallets).length;
	}

	while (profile.wallets().findByAlias(makeAlias(sameCoinWalletsCount))) {
		sameCoinWalletsCount++;
	}

	return makeAlias(sameCoinWalletsCount);
};
