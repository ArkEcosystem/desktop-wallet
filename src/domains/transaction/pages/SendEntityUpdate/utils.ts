import { Contracts } from "@arkecosystem/platform-sdk";
import { File } from "@arkecosystem/platform-sdk-ipfs";
import { Enums, ReadWriteWallet, TransactionData } from "@arkecosystem/platform-sdk-profiles";
import { filter, isEmpty } from "@arkecosystem/utils";
import { httpClient } from "app/services";

type SignEntityUpdateProps = {
	form: any;
	senderWallet: ReadWriteWallet;
	env: any;
	type?: number;
};

export const fetchTxIpfsData = async (tx: TransactionData) => {
	const data = tx.asset().data as { ipfsData: string };
	const ipfsData = await new File(httpClient).get(data.ipfsData);

	if (!ipfsData) throw new Error(`Unable to find Ipfs Data for ${tx.id()}`);
	return ipfsData;
};

export const sendEntityUpdateTransaction = async ({ form, senderWallet, env, type }: SignEntityUpdateProps) => {
	const { fee, ipfsData, mnemonic, secondMnemonic, registrationId } = form.getValues();

	const sanitizedData = filter(ipfsData, (item) => !isEmpty(item));
	const entityType = type ?? Enums.EntityType.Business;

	const transactionId = await senderWallet.transaction().signEntityUpdate({
		fee,
		from: senderWallet.address(),
		sign: { mnemonic, secondMnemonic },
		data: {
			registrationId,
			type: entityType,
			// @TODO: let the user choose what sub-type they wish to use.
			subType: Enums.EntitySubType.None,
			ipfs: await new File(httpClient).upload(sanitizedData),
		},
	});

	const response: Contracts.BroadcastResponse = await senderWallet.transaction().broadcast(transactionId);
	const errorCodes = Object.values(response?.errors || {}).map((e) => e[0]);
	if (errorCodes.length > 0) throw errorCodes[0];

	await env.persist();

	return senderWallet.transaction().transaction(transactionId);
};
