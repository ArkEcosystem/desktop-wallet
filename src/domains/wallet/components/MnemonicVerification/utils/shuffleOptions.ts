import { shuffle, uniq } from "@arkecosystem/utils";

type Props = {
	options: string[];
	value: string;
	limit: number;
};

export function shuffleOptions({ options, value, limit }: Props) {
	const optionsWithoutAnswer = options.filter((item) => item !== value);
	const availableOptions = uniq(shuffle(optionsWithoutAnswer)).slice(0, limit - 1);
	return shuffle([value, ...availableOptions]);
}
