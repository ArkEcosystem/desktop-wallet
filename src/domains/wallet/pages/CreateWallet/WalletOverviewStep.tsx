import { Alert } from "app/components/Alert";
import { Button } from "app/components/Button";
import { Clipboard } from "app/components/Clipboard";
import { Divider } from "app/components/Divider";
import { Header } from "app/components/Header";
import { Icon } from "app/components/Icon";
import { toasts } from "app/services";
import React, { useEffect, useState } from "react";
import { useFormContext } from "react-hook-form";
import { Trans, useTranslation } from "react-i18next";
import { saveFile } from "utils/electron-utils";

import { MnemonicList } from "../../components/MnemonicList";

export const WalletOverviewStep = () => {
	const { getValues, unregister, watch } = useFormContext();

	// getValues does not get the value of `defaultValues` on first render
	const [defaultMnemonic] = useState(() => watch("mnemonic"));
	const mnemonic = getValues("mnemonic") || defaultMnemonic;

	const [defaultWallet] = useState(() => watch("wallet"));
	const wallet = getValues("wallet") || defaultWallet;

	const { t } = useTranslation();

	useEffect(() => {
		unregister("verification");
	}, [unregister]);

	const handleDownload = async () => {
		const fileName = `${wallet.address()}.txt`;

		try {
			const filePath = await saveFile(mnemonic, fileName, {
				filters: { extensions: ["txt"], name: "Text Document" },
				returnBasename: true,
			});

			if (filePath) {
				toasts.success(
					<Trans
						i18nKey="COMMON.SAVE_FILE.SUCCESS"
						values={{ filePath }}
						components={{ bold: <strong /> }}
					/>,
				);
			}
		} catch (error) {
			toasts.error(t("COMMON.SAVE_FILE.ERROR", { error: error.message }));
		}
	};

	return (
		<section data-testid="CreateWallet__WalletOverviewStep" className="space-y-8">
			<Header title={t("WALLETS.PAGE_CREATE_WALLET.PASSPHRASE_STEP.TITLE")} />

			<Alert>{t("WALLETS.PAGE_CREATE_WALLET.PASSPHRASE_STEP.WARNING")}</Alert>
			<MnemonicList mnemonic={mnemonic} />

			<Divider dashed />

			<div className="flex justify-between items-center">
				<div className="space-y-2">
					<span className="text-lg font-semibold">
						{t("WALLETS.PAGE_CREATE_WALLET.PASSPHRASE_STEP.COPY_OR_DOWNLOAD.TITLE")}
					</span>
					<p className="text-sm text-theme-secondary-text">
						{t("WALLETS.PAGE_CREATE_WALLET.PASSPHRASE_STEP.COPY_OR_DOWNLOAD.DESCRIPTION")}
					</p>
				</div>

				<Icon name="FilePassword" className="text-black dark:text-theme-secondary-600" width={40} height={40} />
			</div>

			<div className="flex justify-end space-x-3 w-full">
				<Clipboard variant="button" data={mnemonic} data-testid="CreateWallet__copy">
					<Icon name="Copy" />
					<span>{t("COMMON.COPY")}</span>
				</Clipboard>

				<Button
					data-testid="CreateWallet__download"
					variant="secondary"
					className="flex items-center space-x-2"
					onClick={handleDownload}
				>
					<Icon name="Download" />
					<span>{t("COMMON.DOWNLOAD")}</span>
				</Button>
			</div>

			<Divider dashed />
		</section>
	);
};
