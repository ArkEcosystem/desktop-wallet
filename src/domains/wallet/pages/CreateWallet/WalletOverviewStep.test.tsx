/* eslint-disable @typescript-eslint/require-await */
import { Contracts } from "@arkecosystem/platform-sdk-profiles";
import { act, renderHook } from "@testing-library/react-hooks";
import { toasts } from "app/services";
import electron from "electron";
import React from "react";
import { FormProvider, useForm } from "react-hook-form";
import { env, fireEvent, getDefaultProfileId, MNEMONICS, render, waitFor } from "utils/testing-library";

import { WalletOverviewStep } from "./WalletOverviewStep";

let profile: Contracts.IProfile;

describe("WalletOverviewStep", () => {
	beforeEach(() => {
		profile = env.profiles().findById(getDefaultProfileId());

		for (const wallet of profile.wallets().values()) {
			profile.wallets().forget(wallet.id());
		}
	});

	describe("Render step", () => {
		it("should render", async () => {
			const { result: form } = renderHook(() =>
				useForm({
					defaultValues: {
						mnemonic: MNEMONICS[0],
						wallet: {
							address: () => "address",
						},
					},
				}),
			);

			jest.spyOn(electron.remote.dialog, "showSaveDialog").mockImplementation(() => ({
				filePath: "filePath",
			}));

			const { getByTestId, asFragment } = render(
				<FormProvider {...form.current}>
					<WalletOverviewStep />
				</FormProvider>,
			);

			expect(asFragment()).toMatchSnapshot();

			const writeTextMock = jest.fn();
			const clipboardOriginal = navigator.clipboard;
			// @ts-ignore
			navigator.clipboard = { writeText: writeTextMock };

			act(() => {
				fireEvent.click(getByTestId("CreateWallet__copy"));
			});

			await waitFor(() => expect(writeTextMock).toHaveBeenCalledWith(MNEMONICS[0]));

			// @ts-ignore
			navigator.clipboard = clipboardOriginal;
		});

		it("should show success toast on succesfull download", async () => {
			const { result: form } = renderHook(() =>
				useForm({
					defaultValues: {
						mnemonic: MNEMONICS[0],
						wallet: {
							address: () => "address",
						},
					},
				}),
			);

			jest.spyOn(electron.remote.dialog, "showSaveDialog").mockImplementation(() => ({
				filePath: "filePath",
			}));

			const toastSpy = jest.spyOn(toasts, "success");

			const { getByTestId } = render(
				<FormProvider {...form.current}>
					<WalletOverviewStep />
				</FormProvider>,
			);

			await act(async () => {
				fireEvent.click(getByTestId("CreateWallet__download"));
			});

			expect(toastSpy).toHaveBeenCalled();

			toastSpy.mockRestore();
		});

		it("should not show success toast on cancelled download", async () => {
			const { result: form } = renderHook(() =>
				useForm({
					defaultValues: {
						mnemonic: MNEMONICS[0],
						wallet: {
							address: () => "address",
						},
					},
				}),
			);

			jest.spyOn(electron.remote.dialog, "showSaveDialog").mockImplementation(() => ({
				filePath: undefined,
			}));

			const toastSpy = jest.spyOn(toasts, "success");

			const { getByTestId } = render(
				<FormProvider {...form.current}>
					<WalletOverviewStep />
				</FormProvider>,
			);

			await act(async () => {
				fireEvent.click(getByTestId("CreateWallet__download"));
			});

			expect(toastSpy).not.toHaveBeenCalled();

			toastSpy.mockRestore();
		});

		it("should show error toast on error", async () => {
			const { result: form } = renderHook(() =>
				useForm({
					defaultValues: {
						mnemonic: MNEMONICS[0],
						wallet: {
							address: () => "address",
						},
					},
				}),
			);

			jest.spyOn(electron.remote.dialog, "showSaveDialog").mockImplementation(() => {
				throw new Error("Error");
			});

			const toastSpy = jest.spyOn(toasts, "error");

			const { getByTestId } = render(
				<FormProvider {...form.current}>
					<WalletOverviewStep />
				</FormProvider>,
			);

			await act(async () => {
				fireEvent.click(getByTestId("CreateWallet__download"));
			});

			expect(toastSpy).toHaveBeenCalled();

			toastSpy.mockRestore();
		});
	});
});
