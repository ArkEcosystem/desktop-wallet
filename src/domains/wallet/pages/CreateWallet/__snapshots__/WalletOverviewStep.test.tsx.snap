// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletOverviewStep Render step should render 1`] = `
<DocumentFragment>
  <section
    class="space-y-8"
    data-testid="CreateWallet__WalletOverviewStep"
  >
    <div
      class="flex justify-between items-end bg-theme-background"
    >
      <div
        class="space-y-4"
      >
        <h1
          class="mb-0 text-4xl"
          data-testid="header__title"
        >
          Your Passphrase
        </h1>
      </div>
    </div>
    <div
      class="flex overflow-hidden flex-col space-y-4 rounded-lg p-4 bg-theme-warning-50 dark:bg-theme-secondary-800 sm:space-y-0 sm:space-x-4 sm:flex-row sm:items-center"
    >
      <div
        class="h-11 w-11 flex flex-shrink-0 items-center justify-center rounded-lg text-white bg-theme-warning-600"
      >
        <div
          class="sc-bdfBwQ hteSKx"
          height="1.25em"
          width="1.25em"
        >
          <svg>
            alert-warning.svg
          </svg>
        </div>
      </div>
      <div
        class="flex flex-col"
      >
        <p
          class="text-sm leading-relaxed break-words dark:text-theme-secondary-500"
        >
          This is your 12 word Mnemonic. Your mnemonic serves as your password and allows you to send any funds associated with your address. If lost, this mnemonic cannot be restored by anyone, including the ARK.io team. Please store this mnemonic in a safe location and do not share it with anyone.
        </p>
      </div>
    </div>
    <ul
      class="grid grid-cols-2 gap-x-3 gap-y-5 md:grid-cols-3 lg:grid-cols-4"
    >
      <li
        class="relative py-3 px-3 rounded border border-theme-secondary-300 dark:border-theme-secondary-800"
        data-testid="MnemonicList__item"
      >
        <span
          class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-800"
        >
          1
        </span>
        <span
          class="font-medium"
        >
          test
        </span>
      </li>
      <li
        class="relative py-3 px-3 rounded border border-theme-secondary-300 dark:border-theme-secondary-800"
        data-testid="MnemonicList__item"
      >
        <span
          class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-800"
        >
          2
        </span>
        <span
          class="font-medium"
        >
          mnemonic
        </span>
      </li>
    </ul>
    <div
      class="flex justify-end w-full"
    >
      <div
        class="inline-block cursor-pointer"
        data-testid="clipboard__wrapper"
      >
        <button
          class="Button__StyledButton-mdtvtf-0 iUSBMV"
          data-testid="CreateWallet__copy"
          type="button"
        >
          <div
            class="flex relative items-center space-x-2"
          >
            <div
              class="sc-bdfBwQ fPrsah"
              height="1em"
              width="1em"
            >
              <svg>
                copy.svg
              </svg>
            </div>
            <span>
              Copy
            </span>
          </div>
        </button>
      </div>
    </div>
    <div
      class="Divider-sc-19q4mlo-0 bKsVMa border-theme-secondary-300 dark:border-theme-secondary-800"
      type="horizontal"
    />
    <div
      class="flex justify-between items-center"
    >
      <div
        class="space-y-2"
      >
        <span
          class="text-lg font-semibold"
        >
          Download Your Passphrase
        </span>
        <p
          class="text-sm text-theme-secondary-text"
        >
          You can also download and safely store your passphrase.
        </p>
      </div>
      <div
        class="sc-bdfBwQ fTnhmh text-black dark:text-theme-secondary-600"
        height="40"
        width="40"
      >
        <svg>
          file-password.svg
        </svg>
      </div>
    </div>
    <div
      class="flex justify-end w-full"
    >
      <button
        class="Button__StyledButton-mdtvtf-0 iUSBMV flex items-center space-x-2"
        data-testid="CreateWallet__download"
        type="button"
      >
        <div
          class="flex relative items-center space-x-2"
        >
          <div
            class="sc-bdfBwQ fPrsah"
            height="1em"
            width="1em"
          >
            <svg>
              download.svg
            </svg>
          </div>
          <span>
            Download
          </span>
        </div>
      </button>
    </div>
    <div
      class="Divider-sc-19q4mlo-0 bKsVMa border-theme-secondary-300 dark:border-theme-secondary-800"
      type="horizontal"
    />
  </section>
</DocumentFragment>
`;
